Make file watching compatible with node version 6.x (runs on 0.10 too)

zaggino opened the pull request:
related to https://github.com/adobe/brackets-shell/pull/543

trying to make brackets work with the new version of node in the shell

cc @ficristo (I couldn't commit to your repo that's why I merged your branch here)

Commit:
Use chokidar instead of the fsevents/fsevents_win combo

Use chokidar 1.6.0 installed with npm@3 and npm install --production

Commit:
Add 'depth' parameter

Remove depth, add ignore

ficristo committed:
Add some logs

ficristo committed:
Use the stats from chokidar

ficristo committed:
Prepared preferences for the paths to ignore during watching

Commit:
Make tests pass

zaggino committed:
do not commit node_modules to the project

zaggino committed:
do not ignore stuff by default in the domain

zaggino committed:
propagate ignores from FileSystem.js

zaggino committed:
stats are not always available (deletes)

zaggino committed:
chokidar allows recursive watching in linux

zaggino committed:
watchedRoot.filter should be used to ignore files from watching

optimize

simplify

ingorichter commented:
Okay, I'm able to build on OSX. Brackets launches and changes in the opened folder will show up in the project tree. Adding a file in the finder will update the project tree. The same applies for deleting a file. The whole thing stops working, the moment I open a different folder/project. From that point on I don't receive any change notifications anymore and I'm unable to open a different folder/project.

zaggino commented:
Thanks @ingorichter , same thing happening on Windows, I'll be looking into it when I'll have some free time.

zaggino committed:
fix project switching blocker

cleaning the code a bit

zaggino commented:
@ingorichter opening a different project/folder should be fixed now.

zaggino committed:
do not hide true event names from shell file system

zaggino committed:
correctly transfer stats from watcher domain

update travis node_js

zaggino commented:
Good news, all my testing is passing and I can't see any more problems. Now the question: `chokidar` allows pushing ignore property to the library itself to ignore some entries, which is great.

The way it currently works: brackets-web get all file change notifications (even from .git folders, and that's a heap of notifications) from brackets-shell and these are filtered on the web side.

The way it could work: brackets-web could pass filter to the brackes-shell to use with chokidar and these would be filtered on shell side which would be great (much less change events and load on the domain connection) but it would also mean that the api between web and shell would have to change. This would be concern for other brackets-web implementations which doesn't use brackets-shell.

Question is: make this breaking change in this PR or not?

@ingorichter @ficristo @petetnt @busykai @abose @MarcelGerber

ingorichter commented:
👏  @zaggino. That works way better now. Thanks for resolving this issue.
I like the idea of having everything filtered on the shell side. Perhaps we can have a feature flag to have both available. At least for some time before dropping the current implementation. It will be a great improvement over the existing implementation and should improve the performance (at least a bit). Perhaps we could even get rid of the **TOO_MANY_FILES** limitation.

petetnt commented:
I am with @ingorichter with having it behind a feature flag but enabled by default. It brings huge benefits for Brackets. 

Great job with this PR and the current progress with the Node update too in general, really excited about this one 👍

zaggino committed:
make stuff slightly faster using test instead of match

zaggino committed:
add node side filter to the watched root class

zaggino committed:
pass exclusionGlob to WatchedRoot

zaggino committed:
implemented and tested ignoring

zaggino commented:
@ingorichter @petetnt implemented & tested

I've actually been using this branch and related shell build daily for a few days now "in production" and I haven't encountered any side effects.

Considering this finished, now it needs a review and hopefully a merge soon.

petetnt commented:
Great stuff @zaggino, I'll try this as an daily driver today and try to review the code side asap

zaggino committed:
exclude node_modules too

usePolling on windows

zaggino commented:
Excluding node_modules too, watching over them with chokidar causes problems for npm installations and updates to run in a project folder (ref https://github.com/npm/npm/issues/10826#issuecomment-203438132). It'd be highly unusual workflow working inside a node_modules folder anyway.

ingorichter commented:
Okay, I'm testing it now and report back when I see things behaving weird.

ficristo commented:
> Excluding node_modules

I really hope we'll make these user configurable (followup is fine). I think, for example, we should ignore `bower_components` by default too.

zaggino commented:
Followup is fine, I prefer not to build more features until this is merged.

ficristo commented on src/filesystem/impls/appshell/node/FileWatcherDomain.js:
```diff
@@ -99,7 +99,8 @@ function watchPath(path, ignored) {
             followSymlinks: true,
-            ignored: ignored
+            ignored: ignored,
+            usePolling: process.platform === "win32"
```
This is sad, @zaggino why you need this?

petetnt replied:
There's no `fsevents` on Windows so polling is required, right? That is unless we continue to use `fsevents-win` 🐱

zaggino replied:
sadly, `chokidar` causes watched files to lock-up, can't use `npm install` or `bower install` over a watched directory, will need to investigate more

ficristo replied:
I just saw there were already some issues filed against chokidar and the solution seems to use the `usepolling` option 😞 
It seems to me that even VSCode uses its own watcher system on Windows: https://github.com/Microsoft/vscode-filewatcher-windows

By the way on unix they seems to use these options: https://github.com/Microsoft/vscode/blob/0d658f2c0308c770b90df5a73816f0c045817689/src/vs/workbench/services/files/node/watcher/unix/chokidarWatcherService.ts#L30

zaggino replied:
thanks @ficristo , good pointers, I'll update this PR soon

ficristo commented on .travis.yml:
```diff
@@ -1,7 +1,7 @@
 sudo: false # use container-based Travis infrastructure
 node_js:
-  - "0.10"
+  - "6"
```
I don't think this belongs in this PR. (AFAIK chokidar still support 0.10)

zaggino replied:
This PR is tied with Brackets-Shell PR bringing in node 6 to the shell.

zaggino replied:
Reverted back to `0.10` so we can merge this without shell node 6

ficristo commented on package.json:
```diff
@@ -12,6 +12,9 @@
     },
+    "dependencies": {
+        "chokidar": "1.6.0"
+    },
```
We should start using a npm srinkwrap file asap if we start to use the dependecies fields.

zaggino replied:
Added to this PR

ficristo commented on src/filesystem/FileSystem.js:
```diff
@@ -261,7 +261,8 @@ define(function (require, exports, module) {
             recursiveWatch = impl.recursiveWatch,
-            commandName = shouldWatch ? "watchPath" : "unwatchPath";
+            commandName = shouldWatch ? "watchPath" : "unwatchPath",
+            ignored = watchedRoot.globFilter;
```
I think we should consisten we the name. I see ignored, globFilter, globFilter.
Something like `ignoredFilters` would retain a similar name of chokidar option but a bit more descriptive.

zaggino replied:
`ignored` in chokidar allows you to use multiple things that match `anymatch` which are `Array|String|RegExp|Function`. RegExp and function we have to exclude, we can't transport them over to a domain. So just Array and String remains, it can be either glob string or an array of glob strings. Hence I wouldn't use generic ignore name but something that specifies that glob is expected.

ficristo commented on src/filesystem/FileSystem.js:
```diff
@@ -851,11 +852,19 @@ define(function (require, exports, module) {
+        if (typeof globFilter === "function" && typeof callback !== "function") {
+            callback = globFilter;
+            globFilter = null;
+        }
```
Is this logic for backward compatibility? If so add a comment.
(I always find this kind of things hard to follow...)
Is the second condition necessary?

zaggino replied:
updated the condition so it's hopefully more readable, added a comment

ficristo commented on src/filesystem/FileSystemStats.js:
```diff
@@ -36,7 +36,7 @@ define(function (require, exports, module) {
         this._isFile = isFile;
         this._isDirectory = !isFile;
-        this._mtime = options.mtime;
+        this._mtime = options.mtime instanceof Date ? options.mtime : new Date(options.mtime);
```
Could you add a comment where mtime is not a instanceof Date?

zaggino replied:
added comment

ficristo commented on src/filesystem/impls/appshell/AppshellFileSystem.js:
```diff
@@ -112,27 +99,32 @@ define(function (require, exports, module) {
 
     /**
      * Event handler for the Node fileWatcher domain's change event.
+     * event values from chokidar: "change", "ready", "add", "addDir", "unlink", "unlinkDir"
```
I don't think we should mention chokidar

zaggino replied:
fixed here: https://github.com/adobe/brackets/pull/12647/commits/c9811ce339f24fd36fc1c8264cfc7126010e0194
also simplified the events so we can use vscode's csharp watcher for windows later

ficristo commented on src/filesystem/impls/appshell/AppshellFileSystem.js:
```diff
@@ -536,19 +529,21 @@ define(function (require, exports, module) {
-     * string when the operation is complete.
+     * string when the operation is complete. This function needs to mirror the
+     * signature of watchPath and hence the ignored argument is passed from
+     * FileSystem code
```
Could you put this on a new line? And maybe add one or max two line to say why is necessary?

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/18ee475c8b164776fd9509b36c5191e51539768f

ficristo commented on src/filesystem/impls/appshell/AppshellFileSystem.js:
```diff
@@ -585,11 +580,11 @@ define(function (require, exports, module) {
     /**
      * Indicates whether or not recursive watching notifications are supported
-     * by the watchPath call. Currently, only Darwin supports recursive watching.
+     * by the watchPath call. With chokidar, every platform supports recursive watching.
```
Remove the last sentence?

zaggino replied:
Ok

ficristo commented on src/filesystem/impls/appshell/node/FileWatcherDomain.js:
```diff
@@ -104,41 +65,56 @@ function unwatchPath(path) {
 
 /**
+ * Transform Node's native fs.stats to a format that can be sent through domain
+ * @param {stats} Node's fs.stats result
```
Miss the param name.

zaggino replied:
fixed

ficristo commented on src/filesystem/impls/appshell/node/FileWatcherDomain.js:
```diff
@@ -104,41 +65,56 @@ function unwatchPath(path) {
+/**
  * Watch a file or directory.
  * @param {string} path File or directory to watch.
+ * @param {array} ignored List of File or directory to NOT watch.
```
maybe `List of File or directory to ignore during watching`?

zaggino replied:
Ok

ficristo commented on src/filesystem/impls/appshell/node/FileWatcherDomain.js:
```diff
@@ -104,41 +65,56 @@ function unwatchPath(path) {
+            ignorePermissionErrors: true,
+            followSymlinks: true,
+            ignored: ignored,
+            usePolling: process.platform === "win32"
```
Add a comment why we need this.

zaggino replied:
This will go away shortly, I'll port the vscode's csharp watcher here.
It works flawlessly from my experience using VSCode.

ficristo commented on src/filesystem/impls/appshell/node/FileWatcherDomain.js:
```diff
@@ -104,41 +65,56 @@ function unwatchPath(path) {
+        });
+
+        watcher.on("all", function (event, filename, nodeFsStats) {
+            if (event === "raw" || event === "error" || !filename) {
```
If I'm not mistaken, before when there was a filename an event was always emitted.
Are this change intentional?

zaggino replied:
You don't want to process all `chokidar`'s events, even VSCode doesn't care about them all -> https://github.com/Microsoft/vscode/blob/0d658f2c0308c770b90df5a73816f0c045817689/src/vs/workbench/services/files/node/watcher/unix/chokidarWatcherService.ts#L50-L79

ficristo commented on src/project/ProjectModel.js:
```diff
@@ -59,6 +59,13 @@ define(function (require, exports, module) {
     var _exclusionListRegEx = /\.pyc$|^\.git$|^\.gitmodules$|^\.svn$|^\.DS_Store$|^Thumbs\.db$|^\.hg$|^CVS$|^\.hgtags$|^\.idea$|^\.c9revisions$|^\.SyncArchive$|^\.SyncID$|^\.SyncIgnore$|\~$/;
 
     /**
+     * @public
```
I don't think you need this.

zaggino replied:
It's a brackets default, it should be readable and available to others.

ingorichter replied:
And there should be a preference to easily extend this list. But that's a different story.

ficristo replied:
My point it's I don't see used elsewhere. Since it's the default it can safely removed.

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/640d2835e3a5c5df7ca28e316ad05355a2562e44

ficristo commented on src/project/ProjectModel.js:
```diff
@@ -59,6 +59,13 @@ define(function (require, exports, module) {
+     * Glob definition of files and folders that should be excluded directly
+     * inside node domain watching with chokidar
+     */
+    var exclusionGlob = "**/(.pyc|.git|.gitmodules|.svn|.DS_Store|Thumbs.db|.hg|CVS|.hgtags|.idea|.c9revisions|.SyncArchive|.SyncID|.SyncIgnore|node_modules)";
```
I have already written it, but I think this should be an array. 
And chokidar supports string, path and glob.
Could you add `bower_components` too?
Does the `.git` work with your extension? On VSCode i believe they use `.git/objects`

zaggino replied:
done in https://github.com/adobe/brackets/pull/12647/commits/147b5eb84a894678809bd385e849ac183f339cac
ad `.git`, extension can call (and calls)
```
FileSystem.prototype.watch = function (entry, filter, filterGlobs, callback) {
```
with any arguments (filters) it wants, there's no reason to watch insides of `.git` when no git extension is installed

ficristo commented on src/project/ProjectModel.js:
```diff
@@ -88,16 +95,18 @@ define(function (require, exports, module) {
+        return !(
+            new RegExp("[" + invalidChars + "]+").test(filename) ||
+            _illegalFilenamesRegEx.test(filename)
+        );
```
Could you hoist these in a variable? It's a bit hard to read.

zaggino replied:
I prefer to leave it as is, seems readable enough to me.

ficristo commented on src/project/ProjectModel.js:
```diff
@@ -1346,10 +1355,12 @@ define(function (require, exports, module) {
     exports._addWelcomeProjectPath  = _addWelcomeProjectPath;
     exports._isWelcomeProjectPath   = _isWelcomeProjectPath;
     exports._ensureTrailingSlash    = _ensureTrailingSlash;
+    exports._exclusionListRegEx     = _exclusionListRegEx;
```
It seems to me you aren't using this outside of the file. Remove it?

zaggino replied:
Ok

ficristo commented:
I didn't try the PR but I left some comments regardless.
Have you tryed if the non recursive code still works?

zaggino committed:
added npm-shrinkwrap

zaggino committed:
more readable condition

add note to a FileSystemStats constructor

zaggino committed:
use simplified group of watch events

zaggino committed:
explain better

forgotten arg name in jsdoc

jsdoc comment

do a bit of renaming

remove unnecessary export

refactor FileWatcherDomain

zaggino commented:
Comments addressed, I haven't played with the non recursive stuff as I see little point - the code around `recursiveWatch` being false wasn't touched in this PR.

zaggino committed:
add c# file watcher executables from vscode

zaggino committed:
add useful options to chokidar

zaggino committed:
add glob

zaggino committed:
refactoring

zaggino committed:
implement CSharpWatcher

reverted node in travis

update coment

forgotten export

zaggino commented:
@ficristo ready for round 2, this now supports Code's CSharpWatcher in `win32`

zaggino committed:
avoid circular dependency errors

zaggino committed:
do not notify about terminations when closing watcher

zaggino committed:
optimize: do not call domain function with an empty list

zaggino committed:
ensure _hash is present in file system stats

refactor, fix stats bug

zaggino committed:
fix the bug with shell rounding the stats mtime causing false blind write notifications

ficristo commented:
> this now supports Code's CSharpWatcher in win32

Sorry if I wasn't clear... I didn't imply to move to it. But it seems you think it's fine so I trust you.

zaggino commented:
@ficristo It's a good solution to avoid `usePolling` with chokidar on Windows. Insides probably are similar/same as `fsevents_win` that Brackets has right now. It's a good, working, tested and proven solution (I never had problems with file watching on Code). I think it'd benefit us to use it and don't have to care about maintaining something of our own.

zaggino committed:
add MS license for CodeHelper.exe

fix jsdoc

rename fwm to FileWatcherManager

remove tabs

update comment

add missed param to jsdoc

ficristo commented on src/filesystem/WatchedRoot.js:
```diff
@@ -59,6 +61,11 @@ define(function (require, exports, module) {
     WatchedRoot.prototype.filter = null;
 
     /**
+     * @type {string}
```
Array<string>

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/640d2835e3a5c5df7ca28e316ad05355a2562e44

ficristo commented on src/filesystem/impls/appshell/AppshellFileSystem.js:
```diff
@@ -114,25 +101,31 @@ define(function (require, exports, module) {
-     * @param {string=} filename The name of the file that changed.
+     * @param {string} event The type of the event: "changed", "created" or "deleted"
+     * @param {string} parentDirPath The path to the directory holding entry that has changed
+     * @param {string=} entryName The name of the file/directory that has changed
```
missing statsObj param

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/640d2835e3a5c5df7ca28e316ad05355a2562e44

ficristo commented on src/filesystem/impls/appshell/AppshellFileSystem.js:
```diff
@@ -520,9 +513,10 @@ define(function (require, exports, module) {
      * when the recursiveWatch property of this module is true.
      *
      * @param {string} path
+     * @param {array} ignored
```
`Array<string>`?

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/640d2835e3a5c5df7ca28e316ad05355a2562e44

ficristo commented on src/filesystem/impls/appshell/AppshellFileSystem.js:
```diff
@@ -536,19 +530,21 @@ define(function (require, exports, module) {
      *
      * @param {string} path
      * @param {function(?string)=} callback
      */
```
missing ignored param

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/d8b5a5dbf9cac5a774b65db7652c7a6b9f4d64ce

ficristo commented on src/filesystem/impls/appshell/node/CSharpWatcher.js:
```diff
@@ -0,0 +1,129 @@
+/*
+ * This copies the implementation of VSCode
+ * https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/files/node/watcher/win32/csharpWatcherService.ts
```
Please, do not use the link with blob/master but the one with the commit SHA.
Maybe add the changes you made?

zaggino replied:
I didn't copy this file, I have rewritten it to JavaScript, code uses TypeScript

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/1c8847ee9ea5894c65593f9c6f01b3badfd30c91

ficristo commented on src/filesystem/impls/appshell/node/CSharpWatcher.js:
```diff
@@ -0,0 +1,129 @@
+var fspath = require("path");
+var cp = require("child_process");
+var anymatch = require('anymatch');
+var fwm = require("./FileWatcherManager");
```
Usually we spell out the name `var FileWatcherManager = ...`

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/7ff5d1e2b7268dcc681acbf6f1aab80674927ab9

ficristo commented on src/filesystem/impls/appshell/node/CSharpWatcher.js:
```diff
@@ -0,0 +1,129 @@
+		handle.stderr.on("data", onError);
+
+		// Exit
+		handle.on("exit", onExit);
```
Are these tab character in upstream? Maybe remove them?

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/0fa1c7d4ddc6a8f52a288d9e20a1a9013258c938

ficristo commented on src/filesystem/impls/appshell/node/ChokidarWatcher.js:
```diff
@@ -0,0 +1,59 @@
+
+var fspath = require("path");
+var chokidar = require("chokidar");
+var fwm = require("./FileWatcherManager");
```
Usually we spell out the name: `var fwm = FileWatcherManager`

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/7ff5d1e2b7268dcc681acbf6f1aab80674927ab9

ficristo commented on src/filesystem/impls/appshell/node/FileWatcherManager.js:
```diff
@@ -0,0 +1,108 @@
+}
+
+/**
+ * @private
```
Move the private annotation after the description.

zaggino replied:
https://github.com/adobe/brackets/pull/12647/commits/640d2835e3a5c5df7ca28e316ad05355a2562e44

ficristo commented on src/filesystem/impls/appshell/node/win32/LICENSE:
```diff
@@ -0,0 +1,18 @@
+THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES 
+OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS 
+BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT 
+OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
I don't know, but should we add this license in the NOTICE file at the root of the project too?

zaggino replied:
It's Microsoft license next to the two microsoft files copied from their project, I don't think it needs to be in root.

ficristo commented:
I left some other comments.
Before I forget: AFAIK Brackets build scripts do not take in consideration the package dependencies, we should check this.

zaggino commented:
@ficristo done, hope I didn't miss any comments.

npm dependencies PR opened here: https://github.com/adobe/brackets-shell/pull/572

that's all that should be needed, grunt copies everything, even node_modules from adobe/brackets clone: https://github.com/adobe/brackets-shell/blob/master/Gruntfile.js#L177-L184

ficristo commented:
Two things I almost forgotten:
- before we merge this we should squash the commits. @zaggino you can remove mine: you deserve all the credits!
- chokidar come with an optional dependecy: fsevents. I suppose we build on OSX before releasing the installer for that platform. Right?

zaggino commented:
@ficristo 

a) no commit removing, you laid down a base for this PR and credit is due for that (you also did a good job reviewing it), merge as is

b) you can't build osx shell on non-osx machine so that's right

ficristo commented:
I'm sorry I noticed only now: in newer file you should add the classic license comment of Adobe.
I'm not sure if `CSharpWatcher.js` needs one

ficristo commented:
I don't think the PR for the build system is right: I'll write more there.
I want this ASAP in the core, so I'm fine to fix the build part in a followup.

ficristo commented:
For the commits: I don't pretend you squash everything in a single commit.
But I think commits like these two could be squashed together:
`Add 'depth' parameter` 8ecb4d72af74e4a0138a4f8e9702993955e9828b
`Remove depth, add ignore` 5f91ca19cece64afc5780a5db160e6b8fa938f6e
And probably this could be removed:
`usePolling on windows` e3dbb179026fd9216cdc24db79cbef9d7c74fa68

If you really want to retain all the commits I would like to see better messages.
When I look at the git history, I don't know to what the logs belongs, so I find helpful to see meaningfull logs.
These two messages, for example, don't mean much outside of this PR.
`optimize` 2b4abcfd308563f1b091b5fb4e1ae243eb7b5bc6
`simplify` 020f74b0fd2900ea05163462782ce9d8fe5cab37

Sincerely I'm not sure about the policy of Brackets on these things, so it's up to you what to do.

ficristo commented:
This time I've run, at least, all the test suite on my Windows 10 PC and it looks very green :fireworks:

ficristo commented:
Out of curiosity, how did you generate the `CodeHelper.exe`?

zaggino committed:
add adobe license headers to new files

zaggino commented:
Took it from here https://github.com/Microsoft/vscode/blob/314e122b16c5c1ca0288c8006e9c9c3039a51cd7/src/vs/workbench/services/files/node/watcher/win32/CodeHelper.exe. Also I've tried rebuilding file from the sourcecode's in production mode as advised and got the same filesize (83968 bytes).

zaggino commented:
I'm gonna rebase this and rewrite some commit messages, saying that, there's no policy on this and I've never seen anyone really doing that... maybe taking it a bit too far.

zaggino committed:
add npm install --production script to grunt build

zaggino commented:
@ficristo this should be ready for merging now. I've squashed commits that didn't have much of a meaning on their own like "refactor" or "add comment". Grunt script for installing production node_modules into a dist folder is also present (and tested).

ficristo commented:
I think we had iterate on this PR enough for merging.
@petetnt @ingorichter if you have found any issues we can look in a followup.

@zaggino amazing work. Thank you.
